File: hermes/domain/video_factory.py
"""Domain contracts for the F1-F5 Video Factory."""
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from enum import Enum
from typing import Any
from uuid import uuid4


def new_id(prefix: str) -> str:
    return f"{prefix}_{uuid4().hex[:16]}"


class ClaimStatus(str, Enum):
    VERIFIED = "verified"
    USER_PROVIDED_UNVERIFIED = "user_provided_unverified"
    UNSUPPORTED = "unsupported"
    RESTRICTED = "restricted"


class ProjectStatus(str, Enum):
    DRAFT = "draft"
    RESOURCE_READY = "resource_ready"
    BRIEF_READY = "brief_ready"
    SCENE_PLAN_READY = "scene_plan_ready"
    READY_FOR_STORYBOARD = "ready_for_storyboard"
    STORYBOARD_READY = "storyboard_ready"
    STORYBOARD_APPROVED = "storyboard_approved"
    SCENES_GENERATED = "scenes_generated"
    TIMELINE_READY = "timeline_ready"
    DRAFT_VIDEO_READY = "draft_video_ready"
    READY_TO_PUBLISH = "ready_to_publish"


class FrameGenerationStatus(str, Enum):
    PLANNED = "planned"
    GENERATING = "generating"
    COMPLETED = "completed"
    FAILED = "failed"
    REJECTED = "rejected"


class StoryboardApprovalStatus(str, Enum):
    PENDING = "pending"
    APPROVED = "approved"
    REVISION_REQUIRED = "revision_required"


class VideoGenerationStatus(str, Enum):
    PENDING = "pending"
    GENERATING = "generating"
    COMPLETED = "completed"
    FAILED = "failed"
    REJECTED = "rejected"


class TimelineStatus(str, Enum):
    DRAFT = "draft"
    RENDERING = "rendering"
    COMPLETED = "completed"
    FAILED = "failed"


class FinalApprovalStatus(str, Enum):
    PENDING = "pending"
    APPROVED = "approved"
    REVISION_REQUIRED = "revision_required"


@dataclass(frozen=True)
class AssetReference:
    asset_id: str
    uri: str
    metadata: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if not self.asset_id.strip() or not self.uri.strip():
            raise ValueError("asset_id and uri are required")


@dataclass(frozen=True)
class ResourceIdentity:
    description: str
    shape: str = ""
    color: str = ""
    materials: str = ""
    logo_placement: str = ""
    distinctive_features: tuple[str, ...] = ()

    def __post_init__(self) -> None:
        if not self.description.strip():
            raise ValueError("identity description is required")


@dataclass(frozen=True)
class ResourcePack:
    id: str
    owner_user_id: str
    product_references: tuple[AssetReference, ...]
    primary_product_asset_id: str
    product_identity_description: str
    locked_product_identity: ResourceIdentity | None = None
    character_references: tuple[AssetReference, ...] = ()
    primary_character_asset_id: str | None = None
    character_identity_description: str = ""
    locked_character_identity: ResourceIdentity | None = None
    default_outfit: str = ""
    context: str = ""
    visual_style: str = ""
    locked_at: str | None = None
    version: int = 1

    def __post_init__(self) -> None:
        if not self.owner_user_id.strip() or not self.product_references:
            raise ValueError("owner_user_id and product_references are required")
        product_ids = {item.asset_id for item in self.product_references}
        if self.primary_product_asset_id not in product_ids:
            raise ValueError("primary_product_asset_id must reference a product image")
        if self.primary_character_asset_id and self.primary_character_asset_id not in {
            item.asset_id for item in self.character_references
        }:
            raise ValueError("primary_character_asset_id must reference a character image")
        if self.locked_product_identity and not self.locked_at:
            raise ValueError("locked product identity requires locked_at")


@dataclass(frozen=True)
class RawIdea:
    text: str
    required_elements: tuple[str, ...] = ()
    required_cta: str = ""
    target_duration_seconds: int | None = None
    platform: str = ""
    aspect_ratio: str = ""

    def __post_init__(self) -> None:
        if not self.text.strip():
            raise ValueError("raw idea text is required")
        if self.target_duration_seconds is not None and self.target_duration_seconds <= 0:
            raise ValueError("target_duration_seconds must be positive")


@dataclass(frozen=True)
class Claim:
    claim: str
    status: ClaimStatus
    evidence_refs: tuple[str, ...] = ()
    restriction_reason: str = ""

    def __post_init__(self) -> None:
        if not self.claim.strip():
            raise ValueError("claim is required")
        if self.status in (ClaimStatus.UNSUPPORTED, ClaimStatus.RESTRICTED) and not self.restriction_reason.strip():
            raise ValueError("unsupported or restricted claims require a reason")


@dataclass(frozen=True)
class CreativeBrief:
    objective: str
    target_audience: str
    core_message: str
    tone: str
    pace: str
    cta: str
    content_blocks: tuple[str, ...]
    verified_selling_points: tuple[Claim, ...] = ()
    restrictions: tuple[str, ...] = ()
    required_content: tuple[str, ...] = ()
    platform: str = ""
    aspect_ratio: str = ""
    target_duration_seconds: int | None = None

    def __post_init__(self) -> None:
        for name in ("objective", "target_audience", "core_message"):
            if not getattr(self, name).strip():
                raise ValueError(f"{name} is required")
        if not self.content_blocks:
            raise ValueError("content_blocks must not be empty")
        if self.target_duration_seconds is not None and self.target_duration_seconds <= 0:
            raise ValueError("target_duration_seconds must be positive")


@dataclass(frozen=True)
class Scene:
    scene_id: str
    order: int
    title: str
    objective: str
    content: str
    main_action: str
    duration_seconds: int
    context: str = ""
    camera_intention: str = ""
    start_state: str = ""
    end_state: str = ""
    required_resources: tuple[str, ...] = ()
    notes: str = ""
    status: str = "planned"

    def __post_init__(self) -> None:
        if self.order < 1 or self.duration_seconds <= 0:
            raise ValueError("scene order and duration_seconds must be positive")
        if not self.scene_id.strip() or not self.title.strip():
            raise ValueError("scene_id and title are required")


@dataclass(frozen=True)
class ScenePlan:
    scenes: tuple[Scene, ...]

    def __post_init__(self) -> None:
        if not self.scenes:
            raise ValueError("scene plan must contain at least one scene")
        orders = [scene.order for scene in self.scenes]
        if len(orders) != len(set(orders)) or sorted(orders) != list(range(1, len(orders) + 1)):
            raise ValueError("scene orders must be unique and contiguous")

    @property
    def total_duration_seconds(self) -> int:
        return sum(scene.duration_seconds for scene in self.scenes)


@dataclass(frozen=True)
class FramePrompt:
    positive_prompt: str
    negative_constraints: str = ""
    product_identity_constraints: str = ""
    character_identity_constraints: str = ""
    composition: str = ""
    camera: str = ""
    lighting: str = ""
    environment: str = ""
    action: str = ""
    reference_asset_ids: tuple[str, ...] = ()
    aspect_ratio: str = ""
    provider_options: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if not self.positive_prompt.strip():
            raise ValueError("positive_prompt is required")


@dataclass(frozen=True)
class StoryboardFrame:
    frame_id: str
    scene_id: str
    order: int
    label: str
    purpose: str
    visual_state: str
    subject_action: str
    product_state: str
    character_state: str
    context: str
    camera_intention: str
    required_resource_ids: tuple[str, ...] = ()
    prompt: FramePrompt | None = None
    generation_status: FrameGenerationStatus = FrameGenerationStatus.PLANNED
    generated_asset_id: str | None = None
    generation_job_id: str | None = None
    review_notes: str = ""
    version: int = 1
    created_at: str = ""

    def __post_init__(self) -> None:
        if not self.frame_id.strip() or not self.scene_id.strip():
            raise ValueError("frame_id and scene_id are required")
        if self.order < 1:
            raise ValueError("order must be positive")


@dataclass(frozen=True)
class Storyboard:
    storyboard_id: str
    project_id: str
    frames: tuple[StoryboardFrame, ...]
    approval_status: StoryboardApprovalStatus = StoryboardApprovalStatus.PENDING
    approval_notes: str = ""
    version: int = 1
    created_at: str = ""
    updated_at: str = ""

    def __post_init__(self) -> None:
        if not self.storyboard_id.strip() or not self.project_id.strip():
            raise ValueError("storyboard_id and project_id are required")
        if not self.frames:
            raise ValueError("storyboard must contain at least one frame")


@dataclass(frozen=True)
class VideoPrompt:
    scene_id: str
    duration_seconds: int
    start_visual_state: str
    end_visual_state: str
    subject_action: str
    product_action: str
    camera_movement: str
    camera_framing: str
    environment_motion: str
    motion_constraints: str = ""
    identity_constraints: str = ""
    reference_frame_ids: tuple[str, ...] = ()
    dialogue_or_vo: str = ""
    negative_constraints: str = ""
    provider_options: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if not self.scene_id.strip():
            raise ValueError("scene_id is required")
        if self.duration_seconds <= 0:
            raise ValueError("duration_seconds must be positive")


@dataclass(frozen=True)
class GeneratedScene:
    scene_id: str
    video_prompt: VideoPrompt
    generation_status: VideoGenerationStatus = VideoGenerationStatus.PENDING
    generated_asset_id: str | None = None
    generation_job_id: str | None = None
    provider_operation_id: str | None = None
    review_notes: str = ""
    version: int = 1
    created_at: str = ""
    updated_at: str = ""

    def __post_init__(self) -> None:
        if not self.scene_id.strip():
            raise ValueError("scene_id is required")


@dataclass(frozen=True)
class TimelineClip:
    clip_id: str
    order: int
    source_asset_id: str
    trim_start_seconds: float = 0.0
    trim_end_seconds: float | None = None
    duration_seconds: float = 0.0
    transition: str = ""
    audio_metadata: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if not self.clip_id.strip() or not self.source_asset_id.strip():
            raise ValueError("clip_id and source_asset_id are required")
        if self.order < 1:
            raise ValueError("order must be positive")


@dataclass(frozen=True)
class Timeline:
    timeline_id: str
    project_id: str
    clips: tuple[TimelineClip, ...]
    audio_track_asset_id: str | None = None
    music_asset_id: str | None = None
    status: TimelineStatus = TimelineStatus.DRAFT
    version: int = 1
    created_at: str = ""
    updated_at: str = ""

    def __post_init__(self) -> None:
        if not self.timeline_id.strip() or not self.project_id.strip():
            raise ValueError("timeline_id and project_id are required")
        if not self.clips:
            raise ValueError("timeline must contain at least one clip")
        orders = [clip.order for clip in self.clips]
        if len(orders) != len(set(orders)) or sorted(orders) != list(range(1, len(orders) + 1)):
            raise ValueError("clip orders must be unique and contiguous")


@dataclass(frozen=True)
class VideoFactoryProject:
    id: str
    owner_user_id: str
    status: ProjectStatus = ProjectStatus.DRAFT
    resource_pack: ResourcePack | None = None
    raw_idea: RawIdea | None = None
    creative_brief: CreativeBrief | None = None
    brief_approval: str = "pending"
    scene_plan: ScenePlan | None = None
    scene_plan_approval: str = "pending"
    storyboard: Storyboard | None = None
    generated_scenes: tuple[GeneratedScene, ...] = ()
    timeline: Timeline | None = None
    draft_video_asset_id: str | None = None
    final_video_asset_id: str | None = None
    final_approval: FinalApprovalStatus = FinalApprovalStatus.PENDING
    final_approval_notes: str = ""
    resource_version: int = 0
    idea_version: int = 0
    brief_version: int = 0
    scene_version: int = 0
    storyboard_version: int = 0
    video_generation_version: int = 0
    timeline_version: int = 0
    created_at: str = ""
    updated_at: str = ""

    def __post_init__(self) -> None:
        if not self.id.strip() or not self.owner_user_id.strip():
            raise ValueError("project id and owner_user_id are required")

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

File: apps/web/src/types/videoFactory.ts
export type StageKey =
  | 'resources'
  | 'brief'
  | 'scenes'
  | 'storyboard'
  | 'generation'
  | 'timeline'
  | 'export'

export type StageState =
  | 'completed'
  | 'active'
  | 'running'
  | 'blocked'
  | 'not_started'
  | 'failed'

export interface StageMeta {
  key: StageKey
  label: string
  shortLabel: string
  description: string
  order: number
}

export const CANONICAL_STAGES: StageMeta[] = [
  { key: 'resources', label: '1. Product Resources', shortLabel: 'Resources', description: 'Product identity & locked references', order: 1 },
  { key: 'brief', label: '2. Creative Brief', shortLabel: 'Brief', description: 'Objective, target audience & core message', order: 2 },
  { key: 'scenes', label: '3. Scene Plan', shortLabel: 'Scenes', description: '30s timeline structure & pacing', order: 3 },
  { key: 'storyboard', label: '4. Storyboard Keyframes', shortLabel: 'Storyboard', description: 'Visual keyframe generation & review', order: 4 },
  { key: 'generation', label: '5. Scene Videos', shortLabel: 'Generation', description: 'Video generation per scene & QC', order: 5 },
  { key: 'timeline', label: '6. Timeline & Voiceover', shortLabel: 'Timeline', description: 'TTS, pacing & rough draft assembly', order: 6 },
  { key: 'export', label: '7. Final Export', shortLabel: 'Export', description: 'Full 30s master video & publishing', order: 7 },
]

export interface AssetRef {
  asset_id: string
  uri?: string
  metadata?: {
    physical_hash_filename?: string
    mime_type?: string
    snapshot_id?: string
    resource_pack_lock_id?: string
    [key: string]: any
  }
}

export interface ResourceIdentity {
  description?: string
  color?: string
  distinctive_features?: string[]
}

export interface ResourcePack {
  id: string
  owner_user_id: string
  product_references: AssetRef[]
  primary_product_asset_id: string
  product_identity_description: string
  locked_product_identity?: ResourceIdentity
  locked_at: string
  version: number
}

export interface CreativeBrief {
  objective: string
  target_audience: string
  core_message: string
  content_blocks: string[]
}

export interface Scene {
  scene_id: string
  order: number
  title: string
  purpose: string
  content: string
  visual_style: string
  duration_seconds: number
  setting: string
  camera_movement: string
}

export interface ScenePlan {
  scenes: Scene[]
}

export interface FramePrompt {
  positive_prompt: string
  negative_constraints: string
  product_identity_constraints: string
  character_identity_constraints?: string
  aspect_ratio?: string
  reference_asset_ids?: string[]
  provider_options?: Record<string, any>
}

export interface StoryboardFrame {
  frame_id: string
  scene_id: string
  order: number
  label: string
  purpose: string
  visual_state: string
  subject_action: string
  product_state: string
  context: string
  camera_intention: string
  required_resource_ids: string[]
  prompt: FramePrompt
  generation_status: 'not_started' | 'generating' | 'completed' | 'failed'
  generated_asset_id?: string
  generation_job_id?: string
  review_notes?: string
  version: number
  created_at?: string
}

export interface Storyboard {
  frames: StoryboardFrame[]
}

export interface VideoPrompt {
  scene_id: string
  duration_seconds: number
  start_visual_state: string
  end_visual_state: string
  subject_action: string
  product_action: string
  camera_movement: string
  camera_framing: string
  environment_motion: string
  identity_constraints: string
  reference_frame_ids: string[]
  dialogue_or_vo: string
  negative_constraints: string
  provider_options: Record<string, any>
}

export interface GeneratedScene {
  scene_id: string
  video_prompt: VideoPrompt
  generation_status: 'not_started' | 'generating' | 'completed' | 'failed'
  generated_asset_id?: string
  generation_job_id?: string
  provider_operation_id?: string | null
  review_notes?: string
  version: number
  created_at?: string
  updated_at?: string
}

export interface TimelineClip {
  scene_id: string
  start_time_seconds: number
  duration_seconds: number
  asset_id: string
}

export interface Timeline {
  clips: TimelineClip[]
  voiceover_asset_id?: string
  total_duration_seconds: number
}

export interface HookVariant {
  variant_id: string
  variant_label: string
  hook_angle: string
  creative_brief: CreativeBrief
  scene_plan: ScenePlan
  final_asset_id: string
  export_status: string
  timeline?: Timeline | null
}

export interface ABVariantSet {
  variants: HookVariant[]
  selected_variant_id: string
}

export interface VideoFactoryProject {
  id: string
  owner_user_id: string
  status: string
  resource_pack?: ResourcePack
  raw_idea?: any
  creative_brief?: CreativeBrief
  brief_approval?: 'draft' | 'approved' | 'rejected'
  scene_plan?: ScenePlan
  scene_plan_approval?: 'draft' | 'approved' | 'rejected'
  storyboard?: Storyboard
  generated_scenes?: GeneratedScene[]
  timeline?: Timeline
  draft_video_asset_id?: string
  final_video_asset_id?: string
  final_approval?: string
  final_approval_notes?: string
  ab_variants?: ABVariantSet
  resource_version?: number
  brief_version?: number
  scene_version?: number
  storyboard_version?: number
  video_generation_version?: number
  timeline_version?: number
  created_at?: string
  updated_at?: string
}

export interface ProjectSummary {
  id: string
  status: string
  product_name?: string
  primary_asset_id?: string
  current_stage?: StageKey
  progress_percent?: number
}

export interface Publication {
  id: string
  platform: 'tiktok' | 'youtube_shorts' | 'instagram_reels'
  project_id: string
  status: 'not_published' | 'uploading' | 'processing' | 'published' | 'failed'
  platform_post_id?: string
  caption?: string
  published_at?: string
  last_error?: string
  created_at: string
  updated_at: string
}

export interface TikTokConnectionStatus {
  connected: boolean
  scopes?: string[]
  expires_at?: string
}
